harmony/kernel/src/component.rs

```diff
@@ -253,7 +259,108 @@ impl Thread {
+                                return Err(CapError::InvalidArgument);
+                            }
+                        }
+
+                        let other_frame = other_table.into_raw();
+                        // SAFETY: Whatever is done here can only affect userspace.
+                        unsafe {
+                            if let Some((previous_frame, _)) =
+                                table.map(offset, other_frame, permissions_into(permissions))
+                            {
+                                KPtr::<AnyPageTable>::from_frame_unchecked(KernelFrame::from_raw(
+                                    previous_frame,
+                                ));
+                            }
+                        }
+                        // Note: How do we flush these entries. Probably let userspace handle that? How?
+                        Ok(0)
+                    }
+                    PageTableOp::Unlink { slot } => {
+                        let offset = PageTableOffset::new_truncate(slot as u16);
+                        if flags.level() == 4 {
+                            if !offset.is_lower_half() {
+                                // Trying to modify the higher half kernel address space.
+                                return Err(CapError::InvalidArgument);
+                            }
+                        }
+                        if flags.level() == 1 {
+                            return Err(CapError::InvalidArgument);
+                        }
+                        unsafe {
+                            if let Some((frame, _)) = table.unmap(offset) {
+                                KPtr::<AnyPageTable>::from_frame_unchecked(KernelFrame::from_raw(
+                                    frame,
+                                ));
+                            }
+                        }
+                        Ok(0)
+                    }
+                    PageTableOp::MapFrame {
+                        user_frame,
+                        slot,
+                        permissions,
+                    } => {
+                        let offset = PageTableOffset::new_truncate(slot as u16);
+                        if flags.level() != 1 {
+                            return Err(CapError::InvalidArgument);
+                        }
+
+                        let frame = this
+                            .component()
+                            .user_region_frame(user_frame)?
+                            .try_as_user()?
+                            .into_raw();
+
+                        unsafe {
+                            if let Some((_previous_frame, _)) =
+                                table.map(offset, frame, permissions_into(permissions))
+                            {
+                                // FIXME: This doesn't do any sort of flushing that guarantees the frame has been forgotten!
```

Comment:
Sounds like we need to track flushings per core before releasing a user frame. Once the frame reaches 0 and all pages have been flushed (or the entire tlb was shotdown), then we can release the frame.